src/constraints.jl

```diff
@@ -0,0 +1,559 @@
+    allunique(s) ||_error("Not all the logical indicator variables are unique.")
+    for lvref in s
+        if !JuMP.is_valid(model, lvref)
+            _error("`$lvref` is not a valid logical variable reference.")
+        elseif !haskey(_indicator_to_constraints(model), JuMP.index(lvref))
+            _error("`$lvref` is not associated with any constraints.")
+        end
+    end
+    return s
+end
+
+# fallback
+function _process_structure(_error, s, model::JuMP.Model)
+    _error("Unrecognized disjunction input structure.") # TODO add details on proper syntax
+end
+
+# Write the main function for creating disjunctions that is macro friendly
+function _disjunction(
+    _error::Function,
+    model::JuMP.Model, # TODO: generalize to AbstractModel
+    structure::Vector,
+    name::String
+)
+    is_gdp_model(model) || error("Can only add disjunctions to `GDPModel`s.")
+
+    # build the disjunction
+    indicators = _process_structure(_error, structure, model)
+    disjunction = Disjunction(JuMP.index.(indicators))
+
+    # add it to the model
+    disjunction_data = ConstraintData(disjunction, name)
+    idx = _MOIUC.add_item(_disjunctions(model), disjunction_data)
+
+    _set_ready_to_optimize(model, false)
+    return DisjunctionRef(model, idx)
+end
+
+# Fallback disjunction build for nonvector structure
+function _disjunction(
+    _error::Function,
+    model::JuMP.Model, # TODO: generalize to AbstractModel
+    structure,
+    name::String
+)
+    _error("Unrecognized disjunction input structure.")
+end
+
+# Disjunction build for nested disjunctions with indicator variable given
+function _disjunction(
+    _error::Function,
+    model::JuMP.Model, # TODO: generalize to AbstractModel
+    structure,
+    name::String,
+    tag::DisjunctConstraint
+)
+    dref = _disjunction(_error, model, structure, name)
+    obj = JuMP.constraint_object(dref)
+    return JuMP.add_constraint(model, _DisjunctConstraint(obj, tag.indicator), name)
+end
+
+# General fallback for additional arguments
+function _disjunction(
+    _error::Function,
+    model::JuMP.Model, # TODO: generalize to AbstractModel
+    structure,
+    name::String,
+    extra...
+)
+    for arg in extra
+        _error("Unrecognized argument `$arg`.")
+    end
+end
+
+"""
+    disjunction(
+        model::JuMP.Model, 
+        structure, 
+        name::String = ""; 
+        nested::Bool = false,
+        indicator::LogicalVariableRef
+    )
+
+Function to add a [`Disjunction`](@ref) to a [`GDPModel`](@ref).
+"""
+function disjunction(
+    model::JuMP.Model, 
+    structure, 
+    name::String = ""; 
+    nested::Bool = false,
+    indicator::LogicalVariableRef
+) # TODO add kw argument to build exactly 1 constraint
+    if nested
+        return _disjunction(error, model, structure, name, DisjunctConstraint)
+    else
+        return _disjunction(error, model, structure, name, DisjunctConstraint(indicator))
+    end
+end
```

Comment:
This has a few problems. One is that regardless of `nested` it will always add the `DisjunctConstraint` which is incorrectly specified. Moreover, the indicator keyword argument doesn't have a default value. I would recommend the following two definitions:
```julia
function disjunction(
    model::JuMP.Model, 
    disjunct_indicators, # the docstring should have Vector{LogicalVariableRef} even though we don't explicitly type it this way
    name::String = ""
) # TODO add kw argument to build exactly 1 constraint
    return _disjunction(error, model, structure, name)
end

function disjunction(
    model::JuMP.Model, 
    disjunct_indicators, # the docstring should have Vector{LogicalVariableRef} even though we don't explicitly type it this way
    nested_tag::DisjunctConstraint
    name::String = ""
) # TODO add kw argument to build exactly 1 constraint
    return _disjunction(error, model, structure, name, nested_tag)
end
```

Reply:
I don't think these functions are being used anywhere...

Reply:
They are intended for users to have a non macro option to define a single disjunction.

Reply:
done